Add NuGet package support

johnkord opened the pull request:
This is a joint feature between @radhikaj @CodeMonkeyLeet and me. We've added support for a NuGet package for Windows and enabled the samples tests.

edit: This is no longer a WIP PR; this PR is ready for full review and merge when the community approves of it.

andschwa commented on CMakeLists.txt:
```diff
@@ -25,8 +25,14 @@ endif ()
 
-project("Open Enclave SDK" LANGUAGES C CXX ${OE_ASM})
+project("Open Enclave SDK" LANGUAGES C CXX ${OE_ASM}
+  HOMEPAGE_URL "https://github.com/openenclave/openenclave")
```
I like it 👍

andschwa commented on CMakeLists.txt:
```diff
@@ -216,9 +225,20 @@ endif()
+add_subdirectory(samples)
+
+if (WIN32)
+  install(FILES ./scripts/clangw ./scripts/llvm-arw
```
We set `OE_SCRIPTSDIR` above but aren't using it here (which, I mean, I wouldn't anyway, but what is `OE_SCRIPTSDIR` for?)

andschwa commented on cmake/cpack_settings.cmake:
```diff
@@ -22,4 +24,11 @@ set(CPACK_DEBIAN_OEHOSTVERIFY_PACKAGE_DEPENDS "")
+
+# CPack variables for Nuget packages
+set(CPACK_NUGET_PACKAGE_NAME "open-enclave")
+set(CPACK_NUGET_PACKAGE_AUTHORS "Microsoft, Confidential Computing Consortium")
```
🤔

BRMcLaren replied:
I think there may be a hidden benefit to reversing the authors naming order and I would recommend it, but it is a nit.

If authors are read as "Microsoft AND C3" it sounds like it is being branded by Microsoft and then C3 is adding to it Microsoft's efforts.

However, if it reads the reverse, it reads as if Microsoft is adding to C3's effort which I think might look better on paper as we are not looking to impose Microsoft's brand, per se.

johnkord replied:
Should we just have "Confidential Computing Consortium" for this?

BRMcLaren replied:
I'd like to toss in a vote for calling it C3.

Further, if we were to do both, I would consider mentioning Confidential Computing Consortium before Microsoft.

johnkord replied:
Andy and I talked about it, and we decided to rename this AUTHORS field to "Open Enclave SDK Contributors", because technically that is who will be responsible for the release packages.

andschwa commented on cmake/maybe_build_using_clangw.cmake:
```diff
@@ -45,30 +45,30 @@ function(maybe_build_using_clangw OE_TARGET)
     # Setup library tool variables
     set(CMAKE_C_CREATE_STATIC_LIBRARY 
-        "\"${BASH}\" \"${PROJECT_SOURCE_DIR}/scripts/llvm-arw\" \"qc <TARGET> <OBJECTS>\""
+        "\"${OE_BASH}\" \"${OE_SCRIPTSDIR}/llvm-arw\" \"qc <TARGET> <OBJECTS>\""
```
Oh that's where it's used.

andschwa commented on cmake/openenclave-config.cmake.in:
```diff
@@ -38,6 +50,50 @@ if (UNIX)
+  if (UNIX)
+    find_library(SGX_ENCLAVE_COMMON_LIB NAMES sgx_enclave_common)
+  elseif (WIN32)
+    find_library(SGX_ENCLAVE_COMMON_LIB NAMES sgx_enclave_common
```
Would it make more sense to use one `find_library` call with `PATHS` conditionally defined?

andschwa replied:
(Or even just use `HINTS` and not care about conditionally defining it.)

johnkord replied:
Thanks Andy, nice suggestion, I like this.

andschwa commented on cmake/openenclave-config.cmake.in:
```diff
@@ -38,6 +50,50 @@ if (UNIX)
+  if (NOT SGX_ENCLAVE_COMMON_LIB)
+    message(FATAL_ERROR "-- Looking for sgx_enclave_common library - not found")
+  else ()
+    message("-- Looking for sgx_enclave_common library - found")
```
Any non-fatal error message in the package should probably be `VERBSOE` or `DEBUG`.

andschwa commented on cmake/openenclave-config.cmake.in:
```diff
@@ -38,6 +50,50 @@ if (UNIX)
+    if (UNIX)
+      set_target_properties(openenclave::sgx_enclave_common PROPERTIES IMPORTED_LOCATION ${SGX_ENCLAVE_COMMON_LIB})
+    elseif (WIN32)
+      set_target_properties(openenclave::sgx_enclave_common PROPERTIES IMPORTED_LOCATION $ENV{WINDIR}/System32 IMPORTED_IMPLIB ${SGX_ENCLAVE_COMMON_LIB})
```
I would split this on lines just for readability because of the two variables.

andschwa commented on cmake/openenclave-config.cmake.in:
```diff
@@ -55,6 +111,10 @@ endif ()
 # Include the automatically exported targets.
 include("${CMAKE_CURRENT_LIST_DIR}/openenclave-targets.cmake")
+if (WIN32)
+  include("${CMAKE_CURRENT_LIST_DIR}/maybe_build_using_clangw.cmake")
```
😭

andschwa commented on samples/data-sealing/common/CMakeLists.txt:
```diff
@@ -9,15 +9,27 @@ add_custom_command(OUTPUT datasealing_t.h datasealing_t.c datasealing_args.h
+    #
+    # Alternatively we could use a CMake OBJECT library, but that
+    # requires a newish version of CMake.
+    -Wl,--require-defined=__oe_ecalls_table
```
Since this is the only difference, can we just use a generator expression to make it conditionally off on Windows? It'd be a lot less code.

andschwa replied:
`$<$<PLATFORM_ID:Linux>:-Wl,--require-defined=__oe_ecalls_table>`

johnkord replied:
Nice simplification, thanks Andy! Added to PR

andschwa commented on samples/helloworld/enclave/CMakeLists.txt:
```diff
@@ -1,13 +1,19 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
+# TODO: Where is the best place to set this
```
Set what?

radhikaj replied:
My bad. I deleted the line of code after it but did not delete the comment.

andschwa commented on samples/local_attestation/gen_pubkey_header.sh:
```diff
@@ -18,6 +18,7 @@ EOF
 printf 'static const char OTHER_ENCLAVE_PUBLIC_KEY[] =' >> "$destfile"
 while IFS="" read -r p || [ -n "$p" ]
 do
+	p="$(echo $p | tr -d \"\\\r\")"
```
🤣

andschwa commented on samples/test-samples.cmake:
```diff
@@ -24,7 +24,7 @@ else ()
   # against SGX.
   if (USE_LIBSGX)
-    list(APPEND SAMPLES_LIST local_attestation remote_attestation attested_tls)
+     list(APPEND SAMPLES_LIST local_attestation remote_attestation attested_tls)
```
Noticing lots of unnecessary indentation changes (I feel bad mentioning it, but it seems like an editor setting?)

radhikaj replied:
Yes. I am fixing them. Thanks.

andschwa commented:
@johnkord for a WIP PR you can make it a draft, it's very handy!

jazzybluesea commented:
@johnkord Let me know if you need any help from me on the packaging for the host_verify library.

johnkord commented:
Let's see how this works for all of our configurations :)

bors try

johnkord commented:
bors try

johnkord commented:
bors try

johnkord commented:
bors try

CodeMonkeyLeet commented on samples/local_attestation/gen_pubkey_header.sh:
```diff
@@ -18,6 +18,8 @@ EOF
 printf 'static const char OTHER_ENCLAVE_PUBLIC_KEY[] =' >> "$destfile"
 while IFS="" read -r p || [ -n "$p" ]
 do
+    CR=$(printf "\r")
```
This definitely deserves a comment to explain that it is stripping out carriage return characters generated by OpenSSL on Windows, for compatibility with sample code that expects Linux-style LF-only end-of-lines.

Also in remote_attestation version of this script.

johnkord replied:
Yes, this deserves a comment definitely. Without this change, this script produces an unusable header file.

CodeMonkeyLeet commented on samples/test-samples.cmake:
```diff
@@ -24,14 +24,27 @@ else ()
 
-# The prefix is appended to the value given to DESTDIR, e.g. build/install/opt/openenclave/...
-set(INSTALL_DIR ${BUILD_DIR}/install${PREFIX_DIR})
+if (WIN32)
```
Can we combine this if-else clause with the previous since the two are related to the DESTDIR handling differences anyway?

johnkord replied:
Yes, good point!

CodeMonkeyLeet commented on samples/test-samples.cmake:
```diff
@@ -64,45 +83,50 @@ foreach (SAMPLE ${SAMPLES_LIST})
-      RESULT_VARIABLE TEST_RESULT)
-      if (TEST_RESULT)
-        message(WARNING "Samples test '${SAMPLE}' with pkg-config failed!")
+    if (NOT WIN32)
```
Should this be `if (UNIX)` instead? Here and below on line 102 ... (i.e. is that the only dichotomy of OS types we care about?)

johnkord replied:
Good question. I think we mostly choose to use `if (UNIX)` in our cmake files, so I should keep it consistent

johnkord commented on VERSION:
```diff
@@ -1 +1 @@
-v0.7.x
+v0.7.0
```
Ah yes, I meant to mention something about this. NuGet packages don't like "x" in their version numbers, so I think we should either change our versioning guidelines or translate the x to zero when building a NuGet package. I'm not sure which would be best

johnkord commented on windows/README.md:
```diff
@@ -0,0 +1,20 @@
+Windows
```
Sounds good, thanks!

johnkord commented on CMakeLists.txt:
```diff
@@ -25,8 +25,15 @@ endif ()
 set(PROJECT_VERSION ${OE_VERSION})
+set(OE_SCRIPTSDIR "${PROJECT_SOURCE_DIR}/scripts")
+
+find_program(OE_BASH bash)
```
Let me do a pass through the repo for all-caps BASH references and think about what to do. Andy prefers to use variables or namespaces for commonly used variables/targets, to prevent collision with builds of consumers of our SDK

CodeMonkeyLeet replied:
The variable naming is not the main point I was making though; the `BASH` variable explicitly looks up Git Bash for use with the clangw scripts (see comments on line 137), not the one in Windows. I'm guessing that since everything is compiling and running, `find_program` correctly finds Git Bash as well on Windows. The [CMake rules for it](https://cmake.org/cmake/help/v3.0/command/find_program.html) don't make it obvious that's that case though and, particularly if we're using it for samples, have we checked that it robustly finds the right version of bash on Windows?

If nothing else, lines 136-157 are redundant if we decide that `find_program` is sufficient.

johnkord replied:
Oh I see! Thanks Simon, nice spot. I'll address this soon

johnkord commented on cmake/openenclave-config.cmake.in:
```diff
@@ -38,6 +50,42 @@ if (UNIX)
+
+if (NOT TARGET openenclave::sgx_enclave_common)
+  find_library(SGX_ENCLAVE_COMMON_LIB NAMES sgx_enclave_common
+	       HINTS ${NUGET_PACKAGE_PATH}/EnclaveCommonAPI/lib/native/x64-Release)
```
I previously had it under an `if (WIN32)` block and Andy commented earlier in this PR that it could be factored out using the `HINTS` option

johnkord commented on samples/remote_attestation/common/CMakeLists.txt:
```diff
@@ -1,13 +1,21 @@
 
+add_custom_command(OUTPUT enclave_a_pubkey.h
+  DEPENDS public_key_a ${CMAKE_SOURCE_DIR}/gen_pubkey_header.sh
+  COMMAND ${BASH} ${CMAKE_SOURCE_DIR}/gen_pubkey_header.sh enclave_a_pubkey.h ${CMAKE_BINARY_DIR}/enclave_a/public_a.pem)
```
I think this should be OE_BASH. Thanks for the catch!

johnkord commented:
bors try

johnkord commented on CMakeLists.txt:
```diff
@@ -25,8 +25,15 @@ endif ()
+project("Open Enclave SDK" LANGUAGES C CXX ${OE_ASM}
+  HOMEPAGE_URL "https://github.com/openenclave/openenclave")
 set(PROJECT_VERSION ${OE_VERSION})
+set(OE_SCRIPTSDIR "${PROJECT_SOURCE_DIR}/scripts")
```
Since maybe_build_using_clangw is used during both the normal build of the SDK and by a consumer of the SDK, it will need to know how to find the differently-pathed clangw/llvm-arw scripts. I thought that the easiest way to abstract this was to simply make it be a variable that can be defined differently depending on whether it comes from the build for the SDK or from the SDK package itself.

johnkord commented:
bors try

johnkord commented:
Disabled the newly merged switchless sample on Windows :) Let's try again!

bors try

CodeMonkeyLeet commented on docs/GettingStartedDocs/GettingStarted.Windows.md:
```diff
@@ -180,7 +180,7 @@ Normally this is accessible under the `Visual Studio 2017` folder in the Start M
    cd C:\openenclave
```
@radhikaj, I also just noticed a merge conflict: This version of the doc includes the new dependency on Python 3, but that needs to be propagated to the new docs/GettingStartedDocs/Contributors/WindowsManualInstallPrereqs.md

radhikaj replied:
Looks like @johnkord has already added it to the manual install prereqs

BRMcLaren commented on .jenkins/Jenkinsfile:
```diff
@@ -184,11 +184,10 @@ def win2016LinuxElfBuild(String version, String compiler, String build_type) {
                   bat """
                       vcvars64.bat x64 && \
-                      cmake.exe ${WORKSPACE} -G Ninja -DADD_WINDOWS_ENCLAVE_TESTS=ON -DBUILD_ENCLAVES=OFF -DUSE_LIBSGX=ON -DCMAKE_BUILD_TYPE=${build_type} -DLINUX_BIN_DIR=${WORKSPACE}\\linuxbin\\tests -Wdev && \
+                      cmake.exe ${WORKSPACE} -G Ninja -DADD_WINDOWS_ENCLAVE_TESTS=ON -DBUILD_ENCLAVES=OFF -DUSE_LIBSGX=ON -DCMAKE_BUILD_TYPE=${build_type} -DLINUX_BIN_DIR=${WORKSPACE}\\linuxbin\\tests -DNUGET_PACKAGE_PATH=C:/openenclave/prereqs/nuget -Wdev && \
```
I think it's generally better to not use a hardcoded path.

While this is a Jenkins file, it might be good to get in the habit of doing something such as:
```
DNUGET_PACKAGE_PATH=${OE_DIR}/prereqs/nuget```

johnkord replied:
Yes, this is a good idea. We should parameterize this if possible, and that way someone submitting a sandbox run can set this OE_DIR variable to be whatever they want. I'll create an issue to track this (as it was previously hardcoded as well, and I think can be addressed outside the scope of this PR)

johnkord commented on cmake/openenclave-config.cmake.in:
```diff
@@ -12,6 +12,18 @@ set_and_check(OE_LIBDIR "@PACKAGE_CMAKE_INSTALL_LIBDIR@")
 set_and_check(OE_BINDIR "@PACKAGE_CMAKE_INSTALL_BINDIR@")
 set_and_check(OE_DATADIR "@PACKAGE_CMAKE_INSTALL_DATADIR@")
 set_and_check(OE_INCLUDEDIR "@PACKAGE_CMAKE_INSTALL_INCLUDEDIR@")
+set(OE_SCRIPTSDIR "@PACKAGE_CMAKE_INSTALL_BINDIR@/scripts")
```
I can't use set_and_check (I tried it before I changed it to just set) because this directory won't exist on Linux!

johnkord commented on CMakeLists.txt:
```diff
@@ -216,9 +226,24 @@ endif()
+if (WIN32)
+  install(FILES ./scripts/clangw ./scripts/llvm-arw
+    DESTINATION ${CMAKE_INSTALL_BINDIR}/scripts/)
+  install(FILES ./scripts/install-windows-prereqs.ps1
```
Yes, we should track that we need to sign this. Thanks Simon!

Radhika and I talked about this: we're going to publish an install_oe_sdk-windows.md file that describes how to configure your system, but also gives you the option of running this script with another option to configure your system for SDK consumption. That's because a user will be expected to install 7z, nuget, Visual Studio, OpenSSL, LLVM/clang, the Intel PSW executable, the Intel DCAP executable (and all of its internals, like the nuget packages), and a DCAP client.

I imagine it may be difficult to expect a user to install all of this manually, so I think providing a script to do this will make a user's experience easier. We'll probably just offer another option to this script (probably rename LaunchConfiguration -> Configuration) because the above dependencies are just a subset of the contributor dependencies on Windows.

radhikaj replied:
Yes, we will add install_oe_sdk-windows.md.

johnkord commented on cmake/add_dcap_client_target.cmake:
```diff
@@ -19,7 +19,7 @@ function(add_dcap_client_target TARGET_NAME)
 
     # Define the DCAP provider path
-    set(AZURE_DCAP_QUOTEPROV ${CMAKE_SOURCE_DIR}/prereqs/nuget/Azure.DCAP.Windows/dll/dcap_quoteprov.dll)
+    set(AZURE_DCAP_QUOTEPROV ${NUGET_PACKAGE_PATH}/Azure.DCAP.Windows/dll/dcap_quoteprov.dll)
```
Filed, thanks! #2218

johnkord commented on cmake/package_settings.cmake:
```diff
@@ -49,21 +54,28 @@ install(
   DESTINATION ${CMAKE_INSTALL_LIBDIR}/openenclave/cmake
   FILE openenclave-targets.cmake
   COMPONENT OEHOSTVERIFY)
+if (WIN32)
```
Thanks, this was duplicate code I forgot about. Funny enough I was never warned by CMake about this :)

johnkord commented on docs/GettingStartedDocs/Contributors/WindowsSGX1FLCGettingStarted.md:
```diff
@@ -70,7 +70,7 @@ To build debug enclaves:
 mkdir build\x64-Debug
 cd build\x64-Debug
-cmake -G Ninja -DBUILD_ENCLAVES=1 -DUSE_LIBSGX=1 ..\..
+cmake -G Ninja -DBUILD_ENCLAVES=1 -DUSE_LIBSGX=1 -DNUGET_PACKAGE_PATH=C:/your/path/to/intel_nuget_packages ..\..
```
I find that using forward slashes for paths in DEFINES/Environment Variables is less complex (usually) because there's no chance you need to escape the backslash. Let me fix it all up so all the backslashes here are forward slashes

johnkord commented on host/CMakeLists.txt:
```diff
@@ -338,8 +338,7 @@ if (USE_LIBSGX)
   endif ()
-
-  target_link_libraries(oehost PUBLIC sgx_enclave_common sgx_dcap_ql)
+  target_link_libraries(oehost PUBLIC $<BUILD_INTERFACE:sgx_dcap_ql> $<BUILD_INTERFACE:sgx_enclave_common>)
```
When building the SDK, the two sgx libraries will be considered target dependencies by this link line.
When consuming the SDK (using the cmake configuration export), these two are detected by the openenclave-config.cmake package configuration file.

You'll find a line in openenclave-config.cmake.in with this:
```
target_link_libraries(openenclave::oehost INTERFACE openenclave::sgx_enclave_common openenclave::sgx_dcap_ql)
```

Where the openenclave::sgx_enclave_common & openenclave::sgx_dcap_ql libraries are found using find_library first.

This line solves the main problem I had with this entire PR, trying to get oehost to not reference "sgx_enclave_common" in link lines. Previously because these are `PUBLIC` dependencies of oehost for the cmake configuration export, when linking an executable in the samples with openenclave::oehost, it automatically tried to link against sgx_enclave_common and sgx_dcap_ql, not the namespaced openenclave::sgx_enclave_common and openenclave::sgx_dcap_ql. We can't namespace sgx_enclave_common and sgx_dcap_ql (using `install(TARGETS sgx_enclave_common ...)` )because they are `IMPORTED` libraries, and that isn't legal to do.

I learned way more about cmake than I wanted to know trying to figure out the best way to do this with @andschwa 's guidance.

johnkord commented on samples/data-sealing/common/dispatcher.cpp:
```diff
@@ -161,6 +164,7 @@ int ecall_dispatcher::unseal_data(
     size_t seal_key_size = 0;
     uint8_t* key_info = NULL;
     size_t key_info_size = 0;
+    (void)sealed_data_size;
```
I'll create an issue to resolve this. Fixing unused parameters in samples we can definitely push off into an issue unrelated to NuGet package support :)

As for the casts that were introduced that truncate possibly larger values (due to what I assume is MSVC compiler warnings/errors), that needs to get resolved before this merges because we're changing behavior.

johnkord replied:
Added issue addressing this: #2220

johnkord commented on samples/data-sealing/common/dispatcher.cpp:
```diff
@@ -303,7 +308,7 @@ oe_result_t ecall_dispatcher::get_seal_key_and_prep_sealed_data(
     memset(
         (void*)(padded_data + m_data_size),
-        padded_byte_count,
+        (int)padded_byte_count,
```
Issue filed #2219

johnkord commented on samples/data-sealing/enclave_a_v1/CMakeLists.txt:
```diff
@@ -1,5 +1,14 @@
+  include(maybe_build_using_clangw)
+endif()
+
 add_executable(enclave_a_v1 ecalls.cpp)
```
Oh you mean for the maybe_build_using_clangw having an if(WIN32) inside of it?  That's because we won't be distributing the maybe_build_using_clangw file with the Linux builds, so we can't invoke that function, even if it is a no-op on Linux!

johnkord commented on samples/file-encryptor/enclave/keys.cpp:
```diff
@@ -201,6 +201,7 @@ int ecall_dispatcher::cipher_encryption_key(
     unsigned int output_data_size)
 {
     int ret = 0;
+    (void)output_data_size;
```
Added issue addressing this: #2220

johnkord commented on samples/data-sealing/common/dispatcher.cpp:
```diff
@@ -90,9 +90,9 @@ int ecall_dispatcher::seal_data(
         ENCRYPT_OPERATION,
         m_data,
-        m_data_size,
+        (unsigned int)m_data_size,
```
Fixed, thanks Simon.

johnkord commented on samples/data-sealing/host/host.cpp:
```diff
@@ -8,7 +8,10 @@
 #include <sys/stat.h>
 #include <sys/types.h>
+#ifdef __linux__
 #include <unistd.h>
```
I assume so.  @radhikaj do you know why this was necessary to add?

johnkord commented:
Just rebased this branch on master. I think I've addressed all comments in this PR. There are a few open comments to welcome discussion. Specifically, the RELEASE file and how NuGet handles it is a little awkward. I think we may wish to redesign the VERSION file to be something like:

```
OE_MAJOR=0
OE_MINOR=7
OE_PATCH=0
OE_PRERELEASE=alpha
OE_BUILD=1
```

as per https://semver.org/

Then, we can parse this file to produce a package's version number, depending on the type of package. Different package formats (RPM, DEB, NuGet, etc) expect different version strings, and this kind of VERSION file makes it easier to parse for translation into a full version string for a given package format. I'd like to address this in a future design doc.

johnkord commented:
bors try

radhikaj commented:
bors try

johnkord commented:
Thanks everyone for your reviews & comments!

bors r+
